### PR TITLE
Node implementation

### PR DESCRIPTION
**Pull request by IlyaMoskva**
Issue #30
Core Node logic implemented: request pool, internal processor, entry point.
* API changed: Request class extended to have sender and target address. Address as a parameter is removed from methods have Request as well.
* Graph generator + check. Will be removed from Prod or improved, as system must work with .dot files.
* Added simple implementation classes for Network, Hardware, and Storage interfaces to have it tested.
* DsGraph object implements singleton graph to present a global route table. API is under evaluation.

### CODE REVIEW

**Comment by IlyaMoskva** [comment]
@g4s8 This part of functionality is based on changes made in PR #29, but not blocked. It's better to review in in order #29, #31 .

**Comment by IlyaMoskva** [comment]
@g4s8 ,Please keep in mind, the branch is rebased on top of 25_etc, which was rebased on top of origin/master.

**Comment by IlyaMoskva** [comment]
@stepanvalyavskiy ,please do review ASAP, as this PR is blocking next one to solve issue #33.

**Comment by stepanvalyavskiy** [comment]
> @stepanvalyavskiy ,please do review ASAP, as this PR is blocking next one to solve issue #33.

Sorry, I took some time to learn project & read white paper.

File: src/main/com/cqfn/degitx/simulator/Node.kt [hunk 1]
```diff
@@ -14,17 +12,42 @@ class Node(override var hardware: Hardware,
...
     }
 
+    /**
+     * Node internal logic for Request processing
+     */
+    fun processRequest() {
+        var rq : Request = this.hardware.net.receive().poll()
+        while (rq != null) {
+            if (rq.addressee.equals(this.hardware.net.addr)) {
+                // Process incoming requests to THIS node
```
Comment (stepanvalyavskiy) [inline src/main/com/cqfn/degitx/simulator/Node.kt:36]: @g4s8 do we allows comments inside a method? Or do we follow [qulice](https://www.qulice.com/) approach and disallow comments inside a method and force developers to extract methods in such cases?

File: src/main/com/cqfn/degitx/simulator/Request.kt [hunk 2]
```diff
@@ -2,10 +2,51 @@ package com.cqfn.degitx.simulator
 
 /**
  * Stub for Request sent via Network to Address.
- * Data class.
+ *
  * @param sender Initiator of request.
  * @param id Optional ID of request for cases when response is required. Default value is 0 = No Response Required.
  * @param size Size of body instead of real data. It's enough for Simulation.
  * @param timespent Time spent to process this request. Will go to final log for analysis. Increments all way.
  */
```
Comment (stepanvalyavskiy) [inline src/main/com/cqfn/degitx/simulator/Request.kt:10]: There are some new params, please update comment and probably move it to constructor while it is no more a data class.

File: src/main/com/cqfn/degitx/simulator/Node.kt [hunk 3]
```diff
@@ -14,17 +12,42 @@ class Node(override var hardware: Hardware,
...
+                when (rq.type) {
+                    Request.Type.READ -> rq.timespent += this.hardware.storage.read(rq.size)
+                    Request.Type.WRITE -> rq.timespent += this.hardware.storage.write(rq.size)
+                }
+                // Request is processed successfully, log it.
+                //TODO: Change this log to message send. See issue #6
+                println("Request with id " + rq.id + " processed in " + rq.timespent +
+                        " msec on the Node with address " + this.hardware.net.addr)
+            } else { // Redirect this rq to other node
```
Comment (stepanvalyavskiy) [inline src/main/com/cqfn/degitx/simulator/Node.kt:45]: consider to extract it in redirect func

**Review by stepanvalyavskiy — CHANGES_REQUESTED** [review]
Hi, @IlyaMoskva, please see minor comments.

File: src/main/com/cqfn/degitx/simulator/Node.kt (hunk 1 above)
Comment (IlyaMoskva) [reply src/main/com/cqfn/degitx/simulator/Node.kt:36]: @stepanvalyavskiy ,Yes, you're right. I'll move it to main javadoc.

File: src/main/com/cqfn/degitx/simulator/Request.kt (hunk 2 above)
Comment (IlyaMoskva) [reply src/main/com/cqfn/degitx/simulator/Request.kt:10]: Done.

File: src/main/com/cqfn/degitx/simulator/Node.kt (hunk 3 above)
Comment (IlyaMoskva) [reply src/main/com/cqfn/degitx/simulator/Node.kt:45]: Fixed in PR #34

**Review by stepanvalyavskiy — APPROVED** [review]
@IlyaMoskva, thanks

**Comment by IlyaMoskva** [comment]
Thank you, @stepanvalyavskiy.
@g4s8 ,Please, merge changes as they were reviewed.

**Pull request merged**